earth_enterprise/src/common/SharedString.h

```diff
@@ -0,0 +1,129 @@
+// thousands of duplicates of each string in very large projects.
+//
+// With this class, only an integer identifier is stored in each instance (thus
+// drastically reducing the amount of memory needed in large projects), and
+// that identifier is used to retrieve the string value when it is requested.
+// Since the string-to-identifier ratio is 1:1, equality comparisons are also 
+// much faster as we only need to compare the identifiers.
+class SharedString {
+protected:
+   class RefStorage {
+      private:
+        std::unordered_map<uint32_t, std::string> refFromKeyTable;
+        std::unordered_map<std::string, uint32_t> keyFromRefTable;
+        static uint32_t nextID;
+      public:
+        std::string RefFromKey(const uint32_t &key) {
+					notify(NFY_INFO2, "SharedString::RefStorage::RefFromKey called with %u\n", key);
+          auto refIter = refFromKeyTable.find(key);
+          assert(refIter != refFromKeyTable.end());
+					notify(NFY_INFO2, "In SharedString::RefStorage::RefFromKey, found ref %s\n", refIter->second.c_str());
+          return refIter->second;
+        }
+        uint32_t KeyFromRef(const std::string &ref) {
+					notify(NFY_INFO2, "SharedString::RefStorage::KeyFromRef called with %s\n", ref.c_str());
+          auto keyIter = keyFromRefTable.find(ref);
+          if (keyIter != keyFromRefTable.end()) {
+						notify(NFY_INFO2, "In SharedString::RefStorage::KeyFromRef found key %u\n", keyIter->second);
+            return keyIter->second;
+          }
+          else {
+            // We've never seen this ref before, so make a new key
+            uint32_t key = nextID++; 
+            refFromKeyTable.insert(std::pair<uint32_t, std::string>(key, ref));
+            keyFromRefTable.insert(std::pair<std::string, uint32_t>(ref, key));
+						notify(NFY_INFO2, "Added new entry in RefStorage: ref=%s  key=%u\n", ref.c_str(), key);
+            return key;
+          }
+        }
+ 				RefStorage(){
+	          refFromKeyTable.insert(std::pair<uint32_t, std::string>(0, ""));
+            keyFromRefTable.insert(std::pair<std::string, uint32_t>("", 0));
+				}
+    };
+
+    static RefStorage refStore;
+    uint32_t key;
+
+    friend std::ostream & operator<<(std::ostream &out, const SharedString & ref);
+  public:
+    SharedString(): key(0) {}
+
+    SharedString(const SharedString& str): key(str.key) {
+    }
+
+    SharedString(const std::string& str) {
+        key = refStore.KeyFromRef(str);
+    } 
+
+    SharedString(const char* s) {
+        key = refStore.KeyFromRef(s);
+    }
+
+    bool empty() const {
+        return 0 == key;
+    }
+
+    SharedString & operator=(const std::string &str) {
+        key = refStore.KeyFromRef(str);
+        return *this;
+    }
+
+    operator std::string() const {
+        return refStore.RefFromKey(key);
+    }
+
+    std::string toString() const {
+        return refStore.RefFromKey(key);
+    }
+
+    bool operator<(const SharedString &other) const {

```

Comment:
Maybe add operator> for the sake of consistency

Reply:
It would be more consistent, per se. I still have a preference to only add what is necessary at this point, so I'll probably leave it as is unless it becomes a necessity.